qiskit/pulse/transforms/frames.py

```diff
@@ -0,0 +1,143 @@
+from qiskit.pulse.transforms.canonicalization import block_to_schedule
+from qiskit.pulse.library import Signal
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse import channels as chans, instructions
+from qiskit.pulse.frame import Frame
+
+
+def resolve_frames(
+    schedule: Union[Schedule, ScheduleBlock], frames_config: Dict[Frame, Dict]
+) -> Schedule:
+    """
+    Parse the schedule and replace instructions on Frames by instructions on the
+    appropriate channels.
+
+    Args:
+        schedule: The schedule for which to replace frames with the appropriate
+            channels.
+        frames_config: A dictionary with the frame index as key and the values are
+            a dict which can be used to initialized a ResolvedFrame.
+
+    Returns:
+        new_schedule: A new schedule where frames have been replaced with
+            their corresponding Drive, Control, and/or Measure channels.
+
+    Raises:
+        PulseError: if a frame is not configured.
+    """
+    if frames_config is None or not frames_config:
+        return schedule
+
+    if isinstance(schedule, ScheduleBlock):
+        schedule = block_to_schedule(schedule)
+
+    resolved_frames = {}
+    sample_duration = None
+    for frame, settings in frames_config.items():
+        resolved_frame = ResolvedFrame(frame, **settings)
+
+        # Extract shift and set frame operations from the schedule.
+        resolved_frame.set_frame_instructions(schedule)
+        resolved_frames[frame] = resolved_frame
+        sample_duration = settings["sample_duration"]
+
+    if sample_duration is None:
+        raise PulseError("Frame configuration does not have a sample duration.")
+
+    # Used to keep track of the frequency and phase of the channels
+    channel_trackers = {}
+    for ch in schedule.channels:
+        if isinstance(ch, chans.PulseChannel):
+            channel_trackers[ch] = ChannelTracker(ch, sample_duration)
+
+    sched = Schedule(name=schedule.name, metadata=schedule.metadata)
+
+    for time, inst in schedule.instructions:
+        if isinstance(inst, instructions.Play):
+            chan = inst.channel
+
+            if isinstance(inst.pulse, Signal):
+                frame = inst.pulse.frame
+
+                if frame not in resolved_frames:
+                    raise PulseError(f"{frame} is not configured and cannot " f"be resolved.")
+
+                resolved_frame = resolved_frames[frame]
+
+                frame_freq = resolved_frame.frequency(time)
+                frame_phase = resolved_frame.phase(time)
+
+                # If the frequency and phase of the channel has already been set once in
+                # The past we compute shifts.
+                if channel_trackers[chan].is_initialized():
+                    freq_diff = frame_freq - channel_trackers[chan].frequency(time)
+                    phase_diff = frame_phase - channel_trackers[chan].phase(time)
+
+                    if abs(freq_diff) > 1e-8:
```

Comment:
```suggestion
                    if abs(freq_diff) > 1e-8:
```
Should this tolerance be a configurable argument?

Reply:
It can be made configurable. See here: 9927f80.